src/NServiceBus.Core/MessageSession.cs

```diff
@@ -12,53 +12,60 @@ public MessageSession(RootContext context)
             messageOperations = context.Get<MessageOperations>();
         }
 
-        public Task Send(object message, SendOptions sendOptions, CancellationToken cancellationToken = default)
+        public async Task Send(object message, SendOptions sendOptions, CancellationToken cancellationToken = default)
         {
             Guard.AgainstNull(nameof(message), message);
             Guard.AgainstNull(nameof(sendOptions), sendOptions);
-            return messageOperations.Send(new BehaviorContext(context, cancellationToken), message, sendOptions);
+            using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(context.CancellationToken, cancellationToken);
+            await messageOperations.Send(new BehaviorContext(context, linkedTokenSource.Token), message, sendOptions).ConfigureAwait(false);
```

Comment:
These can't still be `return messageOperations.Whatever(…` without an async state machine?

Reply:
Unfortunately no because then the token source would be disposed before the task is scheduled. I was thinking about introducing a slow and a fast path while doing the changes with something like

```
if(cancellationToken == default)
  return messageOperations.Send(...)
else
  return SomeWrappingMethodThatDoesSendAndAlsoCreateALinkedTokenSource()
```

that would allow getting rid of the token source allocation as well as the state machine for the cases when the token is default, but I felt it might be a bit too involved given that this is not a super duuper in the middle of the pipeline hot path and the message session usage is a "slower"  high-level operation. Happy to make that twist though if you want to see the result, and then we can debate whether it is worth it or not. 

Playing devil's advocate on my thinking one could argue that since all pipeline operations do not provide a cancellation token overload but use the incoming pipeline token it is less likely that someone would pass a non-default token to the message operations (since they are used to not using the token based on our design) and then introducing a slow vs hot path might be beneficial. Thoughts?

Reply:
Ah, I get it now. And I agree that it's not on the hot path and the complexity isn't worth it.